### PR TITLE
Exp init error if base is not an operator

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
Running `op=exp(2, base_op)` (coefficient and base_op argument order backwards) creates an Exp where the base operator is a number and the coefficient is an operator, without raising an error. Subsequently calling methods on the created operator either raise errors that are difficult to understand, or return meaningless results.

**Description of the Change:**
Raise a TypeError on initialization if a `base` is passed to `Exp` and the base is not an `Operator`.

**Benefits:**
Given how exponentials are written, trying to put arguments in the order coefficient, operator is an easy mistake to make. Raising an error at initialization avoids confusing error tracebacks or, in certain cases, returning results in the correct format that do not answer the intended question.

### CODE REVIEW

File: pennylane/ops/op_math/exp.py [hunk 1]
```diff
@@ -133,6 +134,10 @@ class Exp(SymbolicOp):
     control_wires = Wires([])
 
     def __init__(self, base=None, coeff=1, do_queue=True, id=None):
+        if base and not isinstance(base, Operator):
+            raise TypeError(
+                f"base is expected to be type None or Operator, but recieved {type(base)}"
```
Comment (trbromley) [inline pennylane/ops/op_math/exp.py:139]:
```suggestion
                f"base is expected to be type None or Operator, but received {type(base)}"
```

File: pennylane/ops/op_math/exp.py [hunk 2]
```diff
@@ -176,6 +177,11 @@ def _unflatten(cls, data, metadata):
 
     # pylint: disable=too-many-arguments
     def __init__(self, base, coeff=1, num_steps=None, id=None):
+
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:180]:
```suggestion
```

File: pennylane/ops/op_math/exp.py [hunk 3]
```diff
@@ -176,6 +177,10 @@ def _unflatten(cls, data, metadata):
 
     # pylint: disable=too-many-arguments
     def __init__(self, base, coeff=1, num_steps=None, id=None):
+        if not isinstance(base, Operator):
+            raise TypeError(
+                f"base is expected to be type None or Operator, but received {type(base)}"
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:182]:
```suggestion
                f"base is expected to be of type Operator, but received {type(base)}"
```

File: tests/ops/op_math/test_exp.py [hunk 4]
```diff
@@ -99,6 +99,12 @@ class DummyOp(qml.operation.Operator):
         op = constructor(DummyOp(1), 2.312)
         assert op.has_diagonalizing_gates is value
 
+    def test_base_is_not_operator_error(self, constructor):
+        """Test that Exp raises an error if a base is provided that is not an Operator"""
+
+        with pytest.raises(TypeError, match="base is expected to be type None or Operator"):
```
Comment (lillian542) [inline tests/ops/op_math/test_exp.py:105]:
```suggestion
        with pytest.raises(TypeError, match="base is expected to be of type Operator, but received "):
```

File: pennylane/ops/op_math/exp.py [hunk 5]
```diff
@@ -176,6 +177,10 @@ def _unflatten(cls, data, metadata):
 
     # pylint: disable=too-many-arguments
     def __init__(self, base, coeff=1, num_steps=None, id=None):
+        if not isinstance(base, Operator):
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:180]:
```suggestion
        if base is not None and not isinstance(base, Operator):
```

File: pennylane/ops/op_math/exp.py [hunk 6]
```diff
@@ -176,6 +177,10 @@ def _unflatten(cls, data, metadata):
...
+            raise TypeError(
+                f"base is expected to be of type Operator, but received {type(base)}"
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:182]:
```suggestion
                f"base is expected to be of type None or Operator, but received {type(base)}"
```

File: tests/ops/op_math/test_exp.py [hunk 7]
```diff
@@ -99,6 +99,12 @@ class DummyOp(qml.operation.Operator):
         op = constructor(DummyOp(1), 2.312)
         assert op.has_diagonalizing_gates is value
 
+    def test_base_is_not_operator_error(self, constructor):
+        """Test that Exp raises an error if a base is provided that is not an Operator"""
+
+        with pytest.raises(TypeError, match="base is expected to be of type Operator, but received "):
```
Comment (lillian542) [inline tests/ops/op_math/test_exp.py:105]:
```suggestion
        with pytest.raises(TypeError, match="base is expected to be of type None or Operator "):
```

File: pennylane/ops/op_math/exp.py [hunk 8]
```diff
@@ -30,6 +30,7 @@
     Operator,
     OperatorPropertyUndefined,
     Tensor,
+    Operator,
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:33]:
```suggestion
```

File: pennylane/ops/op_math/exp.py [hunk 9]
```diff
@@ -176,6 +176,10 @@ def _unflatten(cls, data, metadata):
 
     # pylint: disable=too-many-arguments
     def __init__(self, base, coeff=1, num_steps=None, id=None):
+        if base is not None and not isinstance(base, Operator):
+            raise TypeError(
+                f"base is expected to be of type None or Operator, but received {type(base)}"
```
Comment (mudit2812) [inline pennylane/ops/op_math/exp.py:181]: Is `base=None` supposed to be valid?

Comment (lillian542) [reply pennylane/ops/op_math/exp.py:181]: It used to be the default, so I didn't want to disallow it, but I do think it's weird.

Comment (timmysilv) [reply pennylane/ops/op_math/exp.py:181]: i'd just remove the None check then, plus `None` isn't an instance of `Operator`

**Review by timmysilv — APPROVED** [review]
changelog isn't super necessary for this but you can add if you wanna. looks good!

happy 1 year to this PR!

File: pennylane/ops/op_math/exp.py (hunk 9 above)
Comment (mudit2812) [reply pennylane/ops/op_math/exp.py:181]: I'm really not a big fan of that `None`. @trbromley Thoughts?

Comment (timmysilv) [reply pennylane/ops/op_math/exp.py:181]: not sure, but it was removed (fairly quietly, it seems) in #3691

Comment (trbromley) [reply pennylane/ops/op_math/exp.py:181]: @albi3ro @mudit2812 do you have any memory of this?

Comment (lillian542) [inline pennylane/ops/op_math/exp.py:181]:
```suggestion
                f"base is expected to be of type Operator, but received {type(base)}"
```

File: tests/ops/op_math/test_exp.py [hunk 10]
```diff
@@ -99,6 +99,12 @@ class DummyOp(qml.operation.Operator):
         op = constructor(DummyOp(1), 2.312)
         assert op.has_diagonalizing_gates is value
 
+    def test_base_is_not_operator_error(self, constructor):
+        """Test that Exp raises an error if a base is provided that is not an Operator"""
+
+        with pytest.raises(TypeError, match="base is expected to be of type None or Operator"):
```
Comment (lillian542) [inline tests/ops/op_math/test_exp.py:105]:
```suggestion
        with pytest.raises(TypeError, match="base is expected to be of type Operator"):
```

File: pennylane/ops/op_math/exp.py [hunk 11]
```diff
@@ -176,6 +176,10 @@ def _unflatten(cls, data, metadata):
 
     # pylint: disable=too-many-arguments
     def __init__(self, base, coeff=1, num_steps=None, id=None):
+        if not isinstance(base, Operator):
+            raise TypeError(
+                f"base is expected to be of type Operator, but received {type(base)}"
+            )
```
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:182]:
```suggestion
            raise TypeError(f"base is expected to be of type Operator, but received {type(base)}")
```

**Comment by lillian542** [comment]
[sc-29465]

**Pull request merged**